// jsonrpc-node/lib/http_server.js
// Generated by CoffeeScript 1.12.7
(function() {
  var BasicAuth, Reply, Server, Session, parser,
    hasProp = {}.hasOwnProperty;

  BasicAuth = require("./http_server_basic_auth");

  parser = require("body-parser");

  Reply = require("./reply");

  Session = (function() {
    function Session(res1) {
      this.res = res1;
    }

    Session.prototype.sendData = function(obj) {
      return this.res.json(obj);
    };

    Session.prototype.sendError = function(id, message) {
      return this.res.status(500).json({
        id: id,
        error: message || ""
      });
    };

    Session.prototype.sendNotification = function(method, params) {
      return this.res.json({
        id: null,
        method: method,
        result: params != null ? params : ""
      });
    };

    Session.prototype.sendReply = function(id, params) {
      return this.res.json({
        id: id,
        result: params != null ? params : ""
      });
    };

    return Session;

  })();

  Server = module.exports = function(opt) {
    var handler;
    handler = function(req, res, next) {
      return handler.handle(req, res, next);
    };
    handler.__proto__ = Server;
    handler.methods = {};
    if (opt != null) {
      handler.register(opt);
    }
    return handler;
  };

  Server.register = function(method, callback) {
    var key, results;
    if (callback == null) {
      if (method instanceof Function) {
        return this.defaultMethod = method;
      } else {
        results = [];
        for (key in method) {
          if (!hasProp.call(method, key)) continue;
          callback = method[key];
          results.push(this.methods[key] = callback);
        }
        return results;
      }
    } else {
      return this.methods[key] = callback;
    }
  };

  Server.handle = function(req, res) {
    if (this.auth != null) {
      return this.auth(req, (function(_this) {
        return function(err, user) {
          if ((err != null) || (user == null)) {
            console.log(req.connection.remoteAddress + " not authorized : " + err);
            return res.status(401).json({
              error: "Unauthorized"
            });
          } else {
            return _this.handleNoAuth(req, res, user);
          }
        };
      })(this));
    } else {
      return this.handleNoAuth(req, res, void 0);
    }
  };

  Server.handleNoAuth = function(req, res, user) {
    var e, reply, request;
    try {
      request = req.body instanceof Object ? req.body : JSON.parse(req.body);
      reply = new Reply(new Session(res), request.id);
      return this.execute(request.method, request.params, reply, user);
    } catch (error) {
      e = error;
      console.warn(req.connection.remoteAddress + " invalid request " + e);
      return res.status(500).json({
        error: "invalid request"
      });
    }
  };

  Server.execute = function(method, params, reply, user) {
    if (this.methods[method] != null) {
      return this.methods[method](params, reply, user);
    } else {
      if (this.defaultMethod != null) {
        return this.defaultMethod(method, params, reply, user);
      } else {
        console.warn(req.connection.remoteAddress + " invalid requested method " + method);
        return reply.error("method " + method + " not found");
      }
    }
  };

  Server.setAuth = function(auth) {
    this.auth = auth;
  };

  Server.setBasicAuth = function(authorize) {
    return this.setAuth(new BasicAuth(authorize));
  };

  Server.listen = function(port, host, callback) {
    this.app = require('express')();
    this.server = require('http').createServer(this.app);
    return this._listen(port, host, callback);
  };

  Server.listenSSL = function(port, host, key, cert, callback) {
    this.app = require("express")();
    this.server = require("https").createServer({
      key: key,
      cert: cert
    }, this.app);
    return this._listen(port, host, callback);
  };

  Server._listen = function(port, host, callback) {
    this.app.use(parser.json());
    this.app.use(this);
    return this.server.listen(port, host, callback);
  };

}).call(this);
